app/services/sign_pagination_service.rb

```diff
@@ -0,0 +1,168 @@
+class SignPaginationService # rubocop:disable Metrics/ClassLength
+  include ActionView::Helpers::TagHelper
+  include ActionView::Helpers::UrlHelper
+
+  ##
+  # @param [Integer] current_page_number
+  # @param [Integer] total_num_results
+  # @param [Hash(String => String)] search_query
+  #
+  def initialize(current_page_number:, total_num_results:, search_query:)
+    @current_page_number = current_page_number
+    @search_query = search_query
+    @total_num_pages = (total_num_results.to_f / Sign::RESULTS_PER_PAGE).ceil
+  end
+
+  ##
+  # @return [ActiveSupport::SafeBuffer] pagination HTML to show on page
+  #
+  def pagination_links_html
+    pagination_links =
+      (1..@total_num_pages)
+      .reject { |page_num| page_num_should_be_removed?(page_num) }
+      .map { |page_num| prepare_for_presentation(page_num) }
+
+    pagination_links
+      .unshift(prev_link)
+      .push(next_link)
+      .join("\n")
+      .html_safe
+  end
+
+  private
+
+  def prepare_for_presentation(page_num)
+    return generate_elipsis               if page_num_should_be_replaced_with_elipsis?(page_num)
+    return wrap_as_current_page(page_num) if current_page?(page_num)
+
+    wrap_with_link(page_num)
+  end
+
+  def wrap_as_current_page(page_num)
+    wrap_in_li(content_tag(:span, page_num, class: 'current a'))
+  end
+
+  def wrap_with_link(page_num)
+    wrap_in_li(
+      link_to(content_tag(:span, page_num),
+              build_search_path_for(page_num))
+    )
+  end
+
+  def wrap_in_li(content)
+    content_tag(:li, content)
+  end
+
+  def generate_elipsis
+    wrap_in_li(content_tag(:span, '...', class: 'a'))
+  end
+
```

Comment:
I left the original pagination logic intact (I just renamed things). It is is still very complex but tbh when I extracted it any further it just made it harder to understand.

Reply:
Way easier to understand